core/threaded/scheduler_GEDF_NP_CI.c

```diff
@@ -0,0 +1,516 @@
+                                // long as the worker thread is idle.
+} _lf_sched_thread_info_t;
+
+/**
+ * @brief Information about worker threads. @see _lf_sched_thread_info_t.
+ * 
+ */
+_lf_sched_thread_info_t* _lf_sched_threads_info;
+
+/**
+ * @brief Number of workers that this scheduler is managing.
+ * 
+ */
+size_t _lf_sched_number_of_workers = 1;
+
+/**
+ * @brief Number of workers that are idle.
+ * 
+ * Adding to/subtracting from this variable must be done atomically.
+ * 
+ */
+volatile size_t _lf_sched_number_of_idle_workers = 0;
+
+/**
+ * @brief Mutex that must be acquired by workers before accessing the executing_q.
+ * 
+ */
+lf_mutex_t _lf_sched_executing_q_mutex;
+
+/////////////////// Scheduler Worker API (private) /////////////////////////
+/**
+ * @brief Distribute 'ready_reaction' to the best idle thread.
+ * 
+ * FIXME
+ *
+ * @param ready_reaction A reaction that is ready to execute.
+ */
+static inline void _lf_sched_distribute_ready_reaction(reaction_t* ready_reaction) {
+    DEBUG_PRINT("Scheduler: Trying to distribute reaction %s.", ready_reaction->name);
+    lf_bool_compare_and_swap(&ready_reaction->status, queued, running);
+    if (pqueue_insert(executing_q, ready_reaction) != 0) {
+        error_print_and_exit("Could not add reaction to the executing queue.");
+    }
+}
+
+/**
+ * Return true if the first reaction has precedence over the second, false otherwise.
+ * @param r1 The first reaction.
+ * @param r2 The second reaction.
+ */
+bool _lf_has_precedence_over(reaction_t* r1, reaction_t* r2) {
+    if (LEVEL(r1->index) < LEVEL(r2->index)
+            && OVERLAPPING(r1->chain_id, r2->chain_id)) {
+        return true;
+    }
+    return false;
+}
+
+
+/**
+ * If the reaction is blocked by a currently executing
+ * reaction, return true. Otherwise, return false.
+ * A reaction blocks the specified reaction if it has a
+ * level less than that of the specified reaction and it also has
+ * an overlapping chain ID, meaning that it is (possibly) upstream
+ * of the specified reaction.
+ * This function assumes the mutex is held because it accesses
+ * the executing_q.
+ * @param reaction The reaction.
+ * @return true if this reaction is blocked, false otherwise.
+ */
+bool _lf_is_blocked_by_executing_or_blocked_reaction(reaction_t* reaction) {
+    if (reaction == NULL) {
+        return false;
+    }
+    // The head of the executing_q has the lowest level of anything
+    // on the queue, and that level is also lower than anything on the
+    // transfer_q (because reactions on the transfer queue are blocked
+    // by reactions on the executing_q). Hence, if the candidate reaction
+    // has a level less than or equal to that of the head of the
+    // executing_q, then it is executable and we don't need to check
+    // the contents of either queue further.
+    if (pqueue_size(executing_q) > 0
+            && reaction->index <= ((reaction_t*) pqueue_peek(executing_q))->index) {
```

Comment:
I think I don't understand your point.

For this particular optimization (introduced in #17), the idea is that if the entire index of the `reaction` at hand is already smaller than the index at the head of the `executing_q`,  there is nothing that can block the `reaction` from executing. The executing_q is sorted such that the smallest deadline and level end up at the top.

Are you suggesting that only the level should be checked?

Reply:
What if the first reaction in the executing queue has a later deadline but an earlier level than `reaction`? To me, it looked like this expression would evaluate to `true` in that case because the deadline is in a more significant bit position, and the entire function call would return `false`. But it should return `true`, because the reaction in the executing queue that is at an earlier level might be blocking `reaction`.

I just brought this up to be extra sure, since I'm not sure if we have a lot of test cases involving multiple reactions with different deadlines.